File: lib/Match.py
# coding: latin1
from . import sfm_glob
from . import helpers
import random
import operator
from .Player import Player


class Match(object):

    def __str__(self):
        minutes = str(self.minutes) + "'" + '\t' + self.teams[0].name + '\t ' + str(self.score[0]) + ' x ' + str(self.score[1]) + '\t' + self.teams[1].name

        poss = self.ball_possession()
        last5 = self.ball_possession_last_5_minutes()

        possession = '\t' + 'Possession:' + '\t' + str(poss[0]) + ' x ' + str(poss[1])

        last_5 = '\t' + 'Last 5min:' + '\t' + str(last5[0]) + ' x ' + str(last5[1])

        goalscorers = 'Goals:\n'
        for goalscorer in self.goalscorers:
            spacing = '\t'
            if goalscorer['team'] == self.teams[1]:
                spacing = '\t\t\t   '
            goalscorers += spacing + goalscorer['player'].name + '\t' + str(goalscorer['minute']) + "'\n"

        end = '----------'

        return minutes + '\n' + possession + '\n' + last_5 + '\n' + goalscorers + '\n' + end

    def allow_substitution(self, team):
        if team == self.teams[0]:
            if self.substitutions[0] < 3 and self.minutes > 0:
                return True
        else:
            if self.substitutions[1] < 3 and self.minutes > 0:
                return True
        return False

    def substitution_made_by_team(self, team):
        if team == self.teams[0]:
            self.substitutions[0] += 1
        else:
            self.substitutions[1] += 1

    def print_result(self):
        if self.finished:
            result = self.teams[0].name + '\t ' + str(self.score[0]) + ' x ' + str(self.score[1]) + '\t' + self.teams[1].name
        else:
            result = self.teams[0].name + '\t' + 'x' + '\t' + self.teams[1].name
        print(result)

    def simulate(self, minutes = None):
        if self.teams[0] is None:
            self.end()
            return True
        if self.teams[1] is None:
            self.end()
            return True
        if self.finished:
            self.end()
            return True

        if minutes is None or minutes >= 90:
            while self.minute():
                pass
        else:
            while self.minutes < minutes:
                self.minute()

        self.end()

    def goal_last_minute(self):
        goals_last_minute = [g for g in self.goalscorers if g['minute'] == self.minutes]
        if len(goals_last_minute) > 0:
            return True
        return False

    def choose_goal_scorer(self, team_id):
        choices = []

        if self.teams[team_id].human:
            tits = [p for p in self.teams[team_id].players if p.playing_status == 0]
            for player in tits:
                probability = sfm_glob.MATCH['GOAL PROB PER POSITION'][player.position]
                choices.append((player, probability))

            return helpers.weighted_choice(choices)
        else:
            players = [p['player'] for p in self.goalscorers if p['team'] == self.teams[team_id]]
            if players is None:
                players = []

            for player in players:
                probability = 0.25
                choices.append((player, probability))

            for i in range(len(players) + 1):
                new_player = Player(country = self.teams[team_id].country, skill = 0)
                choices.append((new_player, 0.25))

            return helpers.weighted_choice(choices)

    def goal(self, team_id):
        self.score[team_id] += 1
        goalscorer = self.choose_goal_scorer(team_id)
        self.goalscorers.append({'player' : goalscorer, 'team' : self.teams[team_id], 'minute' : self.minutes})

    def minute(self):
        self.injured_player_out = None

        if self.teams[0] is None:
            self.end()
            return False
        if self.teams[1] is None:
            self.end()
            return False
        if self.finished:
            return False
        if self.minutes >= 90:
            self.end()
            return False

        self.minutes += 1
        
        for team in self.teams:
            for player in team.players:
                if player.playing_status == 0:
                    player.match_minutes += 1
        team_0_skills = self.teams[0].tactical_skill(match = True, minutes = self.minutes)
        team_1_skills = self.teams[1].tactical_skill(match = True, minutes = self.minutes)

        if not self.is_neutral_field:
            team_0_skills = [x * sfm_glob.MATCH['HOME_ADVANTAGE'] for x in team_0_skills]

        # Probability of team 0 having possession
        team_0_attack_prob = helpers.min_max(helpers.balance(team_0_skills[1], team_1_skills[1]), 1 - sfm_glob.MATCH['MAX_POSS'], sfm_glob.MATCH['MAX_POSS'])

        if random.random() <= team_0_attack_prob:
            # attack team 0
            possession = 0
            SKILL_BAL = helpers.balance(team_0_skills[2], team_1_skills[0])
            goal_prob = helpers.min_max(SKILL_BAL, sfm_glob.MATCH['MIN_SKILL_BALANCE'], 1) * sfm_glob.MATCH['MAX_GOAL_PROB_PER_POSS']
        else:
            # attack team 1
            possession = 1
            SKILL_BAL = helpers.balance(team_1_skills[2], team_0_skills[0])
            goal_prob = helpers.min_max(SKILL_BAL, sfm_glob.MATCH['MIN_SKILL_BALANCE'], 1) * sfm_glob.MATCH['MAX_GOAL_PROB_PER_POSS']

        if random.random() <= goal_prob:
            self.goal(team_id = possession)
        else:
            for team in self.teams:
                if team.human:
                    if random.random() <= sfm_glob.MATCH['INJURY_PROBABILITY_PER_MINUTE']: #* (self.minutes / 90):
                        self.injured_player_out = self.player_injured(team)
                        if self.injured_player_out is not None:
                            self.injured_player_out.set_injury()                        

        if self.teams[0].human:
            tits = [p for p in self.teams[0].players if p.playing_status == 0]
            if len(tits) < sfm_glob.MATCH['MINIMUM_PLAYERS']:
                self.score[0] = 0
                self.score[1] = max(3, self.score[1])
                self.end()
                return False

        if self.teams[1].human:
            tits = [p for p in self.teams[1].players if p.playing_status == 0]
            if len(tits) < sfm_glob.MATCH['MINIMUM_PLAYERS']:
                self.score[1] = 0
                self.score[0] = max(3, self.score[1])
                self.end()
                return False

        self.possession[possession] += 1
        self.possession_last_5_minutes.append(possession)
        if len(self.possession_last_5_minutes) > 5:
            self.possession_last_5_minutes.pop(0)
        return True

    def ball_possession(self):
        if self.minutes != 0:
            poss0 = int(round((self.possession[0] / float(self.minutes)) * 100, 0))
            poss1 = 100 - poss0
            if poss0 >= 80:
                poss0 = 80
                poss1 = 20
            if poss1 >= 80:
                poss0 = 20
                poss1 = 80
            return [poss0, poss1]
        else:
            return [50, 50]

    def ball_possession_last_5_minutes(self):
        if len(self.possession_last_5_minutes) > 0:
            poss1 = int(sum(self.possession_last_5_minutes) * (100 / float(len(self.possession_last_5_minutes))))
            poss0 = 100 - poss1
            if poss0 >= 80:
                poss0 = 80
                poss1 = 20
            if poss1 >= 80:
                poss0 = 20
                poss1 = 80
            return [poss0, poss1]
        else:
            return [50, 50]

    def end(self):
        if not self.finished:
            self.teams[0].update_stats_post_match(self.score[0], self.score[1])
            self.teams[1].update_stats_post_match(self.score[1], self.score[0])
            self.minutes = 90
            self.finished = True

    def player_injured(self, team):
        choices = []

        if team.human:
            tits = [p for p in team.players if p.playing_status == 0]
            for player in tits:
                if player.position == 0:
                    probability = sfm_glob.PLAYER['INJURY_PROB_GK']
                else:
                    probability = sfm_glob.PLAYER['INJURY_PROB_NOT_GK']
                choices.append((player, probability))
            return helpers.weighted_choice(choices)
        else:
            return None

    def winner(self):
        if self.score[0] > self.score[1]:
            return self.teams[0]
        elif self.score[1] > self.score[0]:
            return self.teams[1]
        else:
            return None

    def loser(self):
        if self.score[0] < self.score[1]:
            return self.teams[0]
        elif self.score[1] < self.score[0]:
            return self.teams[1]
        else:
            return None

    def __init__(self, teams, minutes = None, score = None, possession = None, possession_last_5_minutes = None, tactical_changes = None, finished = False, substitutions = None, goalscorers = None, injured_player_out = None, is_neutral_field = False):
        self.teams = teams

        if minutes is None:
            minutes = 0
        self.minutes = minutes

        if score is None:
            score = [0, 0]
        self.score = score

        if possession is None:
            possession = [0, 0]
        self.possession = possession

        if possession_last_5_minutes is None:
            possession_last_5_minutes = []
        self.possession_last_5_minutes = possession_last_5_minutes

        if tactical_changes is None:
            tactical_changes = [0, 0]
        self.tactical_changes = tactical_changes

        self.finished = finished

        if substitutions is None:
            substitutions = [0, 0]
        self.substitutions = substitutions

        if goalscorers is None:
            goalscorers = []
        self.goalscorers = goalscorers

        self.injured_player_out = injured_player_out

        self.is_neutral_field = is_neutral_field

File: webapp/src/game/match.ts
import { MATCH, PLAYER } from './constants';
import { balance, minMax, weightedChoice } from './helpers';
import { Player } from './player';
import type { Team } from './team';

export interface GoalRecord {
  player: Player;
  team: Team;
  minute: number;
}

export class Match {
  minutes: number;
  score: [number, number];
  possession: [number, number];
  possessionLast5Minutes: number[];
  tacticalChanges: [number, number];
  finished: boolean;
  substitutions: [number, number];
  goalscorers: GoalRecord[];
  injuredPlayerOut: Player | null;

  constructor(
    public teams: [Team, Team],
    minutes = 0,
    score: [number, number] = [0, 0],
    possession: [number, number] = [0, 0],
    possessionLast5Minutes: number[] = [],
    tacticalChanges: [number, number] = [0, 0],
    finished = false,
    substitutions: [number, number] = [0, 0],
    goalscorers: GoalRecord[] = [],
    injuredPlayerOut: Player | null = null,
    public isNeutralField = false
  ) {
    this.minutes = minutes;
    this.score = score;
    this.possession = possession;
    this.possessionLast5Minutes = possessionLast5Minutes;
    this.tacticalChanges = tacticalChanges;
    this.finished = finished;
    this.substitutions = substitutions;
    this.goalscorers = goalscorers;
    this.injuredPlayerOut = injuredPlayerOut;
  }

  allowSubstitution(team: Team): boolean {
    if (team === this.teams[0]) {
      return this.substitutions[0] < 3 && this.minutes > 0;
    }
    return this.substitutions[1] < 3 && this.minutes > 0;
  }

  substitutionMadeByTeam(team: Team): void {
    if (team === this.teams[0]) {
      this.substitutions[0] += 1;
    } else {
      this.substitutions[1] += 1;
    }
  }

  simulate(limitMinutes?: number): boolean {
    if (!this.teams[0] || !this.teams[1]) {
      this.end();
      return true;
    }
    if (this.finished) {
      this.end();
      return true;
    }
    if (limitMinutes === undefined || limitMinutes >= 90) {
      while (this.minute()) {
        // simulate full match
      }
    } else {
      while (this.minutes < limitMinutes) {
        if (!this.minute()) {
          break;
        }
      }
    }
    this.end();
    return true;
  }

  goal(teamId: 0 | 1): void {
    this.score[teamId] += 1;
    const goalscorer = this.chooseGoalScorer(teamId);
    this.goalscorers.push({ player: goalscorer, team: this.teams[teamId], minute: this.minutes });
  }

  chooseGoalScorer(teamId: 0 | 1): Player {
    const choices: Array<[Player, number]> = [];
    const team = this.teams[teamId];
    if (team.human) {
      const starters = team.players.filter((player) => player.playingStatus === 0);
      starters.forEach((player) => {
        const probability = MATCH['GOAL PROB PER POSITION'][player.position];
        choices.push([player, probability]);
      });
      return weightedChoice(choices);
    }

    const existing = this.goalscorers.filter((goal) => goal.team === team).map((goal) => goal.player);
    existing.forEach((player) => {
      choices.push([player, 0.25]);
    });
    const needed = Math.max(1, 1 + existing.length);
    for (let i = 0; i < needed; i += 1) {
      const newPlayer = new Player({ country: team.country, skill: 0 });
      choices.push([newPlayer, 0.25]);
    }
    return weightedChoice(choices);
  }

  minute(): boolean {
    this.injuredPlayerOut = null;
    if (this.finished || this.minutes >= 90) {
      this.end();
      return false;
    }

    this.minutes += 1;
    this.teams.forEach((team) => {
      team.players.forEach((player) => {
        if (player.playingStatus === 0) {
          player.matchMinutes += 1;
        }
      });
    });

    let team0Skills = this.teams[0].tacticalSkill(true, this.minutes);
    let team1Skills = this.teams[1].tacticalSkill(true, this.minutes);

    if (!this.isNeutralField && this.teams[0].human) {
      team0Skills = team0Skills.map((skill) => skill * MATCH['HOME_ADVANTAGE']);
    }

    const attackProb = minMax(
      balance(team0Skills[1], team1Skills[1]),
      1 - MATCH['MAX_POSS'],
      MATCH['MAX_POSS']
    );

    let possessionSide: 0 | 1;
    let skillBalance: number;
    if (Math.random() <= attackProb) {
      possessionSide = 0;
      skillBalance = balance(team0Skills[2], team1Skills[0]);
    } else {
      possessionSide = 1;
      skillBalance = balance(team1Skills[2], team0Skills[0]);
    }

    const goalProb = minMax(skillBalance, MATCH['MIN_SKILL_BALANCE'], 1) * MATCH['MAX_GOAL_PROB_PER_POSS'];
    if (Math.random() <= goalProb) {
      this.goal(possessionSide);
    } else {
      this.teams.forEach((team) => {
        if (team.human) {
          const injuryProb = MATCH['INJURY_PROBABILITY_PER_MINUTE'] * (this.minutes / 90);
          if (Math.random() <= injuryProb) {
            const injured = this.playerInjured(team);
            if (injured) {
              injured.setInjury();
              this.injuredPlayerOut = injured;
            }
          }
        }
      });
    }

    for (let index = 0; index < this.teams.length; index += 1) {
      const team = this.teams[index];
      if (team.human) {
        const starters = team.players.filter((player) => player.playingStatus === 0);
        if (starters.length < MATCH['MINIMUM_PLAYERS']) {
          this.score[index as 0 | 1] = 0;
          const other = (index === 0 ? 1 : 0) as 0 | 1;
          this.score[other] = Math.max(3, this.score[other]);
          this.end();
          return false;
        }
      }
    }

    this.possession[possessionSide] += 1;
    this.possessionLast5Minutes.push(possessionSide);
    if (this.possessionLast5Minutes.length > 5) {
      this.possessionLast5Minutes.shift();
    }
    return true;
  }

  ballPossession(): [number, number] {
    if (this.minutes === 0) {
      return [50, 50];
    }
    let poss0 = Math.round((this.possession[0] / this.minutes) * 100);
    let poss1 = 100 - poss0;
    if (poss0 >= 80) {
      poss0 = 80;
      poss1 = 20;
    }
    if (poss1 >= 80) {
      poss0 = 20;
      poss1 = 80;
    }
    return [poss0, poss1];
  }

  ballPossessionLast5Minutes(): [number, number] {
    if (!this.possessionLast5Minutes.length) {
      return [50, 50];
    }
    let poss1 = Math.round(
      (this.possessionLast5Minutes.reduce((acc, val) => acc + val, 0) / this.possessionLast5Minutes.length) * 100
    );
    let poss0 = 100 - poss1;
    if (poss0 >= 80) {
      poss0 = 80;
      poss1 = 20;
    }
    if (poss1 >= 80) {
      poss0 = 20;
      poss1 = 80;
    }
    return [poss0, poss1];
  }

  end(): void {
    if (!this.finished) {
      this.teams[0].updateStatsPostMatch(this.score[0], this.score[1]);
      this.teams[1].updateStatsPostMatch(this.score[1], this.score[0]);
      this.minutes = 90;
      this.finished = true;
    }
  }

  playerInjured(team: Team): Player | null {
    if (!team.human) {
      return null;
    }
    const choices: Array<[Player, number]> = [];
    const starters = team.players.filter((player) => player.playingStatus === 0);
    starters.forEach((player) => {
      const probability = player.position === 0 ? PLAYER['INJURY_PROB_GK'] : PLAYER['INJURY_PROB_NOT_GK'];
      choices.push([player, probability]);
    });
    if (!choices.length) {
      return null;
    }
    return weightedChoice(choices);
  }

  goalLastMinute(): boolean {
    return this.goalscorers.some((goal) => goal.minute === this.minutes);
  }

  winner(): Team | null {
    if (this.score[0] > this.score[1]) {
      return this.teams[0];
    }
    if (this.score[1] > this.score[0]) {
      return this.teams[1];
    }
    return null;
  }

  loser(): Team | null {
    if (this.score[0] < this.score[1]) {
      return this.teams[0];
    }
    if (this.score[1] < this.score[0]) {
      return this.teams[1];
    }
    return null;
  }
}
